node/src/actors/chain_manager/mining.rs

```diff
@@ -380,7 +385,44 @@ impl ChainManager {
                         }
                     }
                 })
-                .and_then(move |vrf_proof| {
+                .into_actor(self)
+                // Get collateral
+                .and_then(move |vrf_proof, act, _| {
+                    let coinage = act.chain_state.chain_info.as_ref().unwrap().consensus_constants.collateral_age;
+                    let block_number_limit = act.chain_state.block_number().saturating_sub(coinage);
+                    // Check if we have enough collateral before starting retrieval
+                    match build_commit_collateral(
+                        collateral_amount,
+                        &mut act.chain_state.own_utxos,
+                        own_pkh,
+                        &act.chain_state.unspent_outputs_pool,
+                        timestamp,
+                        tx_pending_timeout,
+                        // The block number must be lower than this limit
+                        block_number_limit
+                    ) {
+                        Ok(collateral) => actix::fut::ok((vrf_proof, collateral)),
+                        Err(TransactionError::NoMoney {
+                                available_balance, transaction_value, ..
+                        }) => {
+                            let required_collateral = transaction_value;
+                            log::warn!("Not enough mature UTXOs for collateral for data request {}: Available balance: {}, Required collateral: {}",
+                                Yellow.bold().paint(dr_pointer.to_string()),
+                                available_balance,
+                                required_collateral,
+                            );
+                            // Decrease the retrieval limit hoping that some other, cheaper,
+                            // data request can be resolved instead
+                            cloned_retrieval_count2.fetch_sub(added_retrieval_count, atomic::Ordering::Relaxed);
```

Comment:
This is interesting and well thought.

Were you able to see this actually happen?

Reply:
> Were you able to see this actually happen?

No